### DISCUSSION TITLE
Phase 2 negotiation fails on FortiGate VPN: INFORMATIONAL_V1 request meets with DELETE for IKE_SA response

### DISCUSSION THREAD

**Discussion by bernd-wechner** [strongswan/strongswan#2343] [Q&A] [answered]
Am using: strongSwan 5.9.5 swanctl
On a Linux Mint 21.3 box.

Trying to connect to my work VPN, which has gone IPSEC.

It's using IKEv1 (alas won't do IKEv2) and I have a successful phase 1 negotiation and IKE_SA. 

I am running on the assumption that what Fortigate call Phase 2, strongswan calls a CHILD_SA.

I am provided this Phase config as guidance:
![image](https://github.com/user-attachments/assets/62cb0818-c299-4c95-94b5-df12e2233db0)

I am using this `swanctl.conf`

```
connections {
    workplace {
        version = 1
        local_addrs = %any
        remote_addrs = fw.echoview.com
        proposals = "aes256-sha256-modp2048"
        aggressive = yes
        send_cert = always

        local-pubkey {
            round = 1
            certs = my-cert.pem
            auth = pubkey
        }

        local-xauth {
            round = 2
            auth = xauth
            xauth_id = my_userid
        }

        remote {
            auth = pubkey
            id = "work_cert_summary"
            cacerts = work.pem
        }

        children {
            vpn {
                mode = tunnel
                esp_proposals = "aes256-sha256-modp2048"
                local_ts  = 0.0.0.0/0
                remote_ts  = 192.168.3.1/32
                start_action = start
            }
        }
    }
}

secrets {
    xauth {
        id = my_userid
        secret = my_password
    }    
}
```

And here's the trace of a connection attempt:

```
$ sudo swanctl --initiate --child vpn --debug 1
[IKE] sending cert request for "DC=au, DC=com, DC=workplace, CN=workplace-servername-CA"
[IKE] initiating Aggressive Mode IKE_SA workplace[3784] to 101.178.197.207
[ENC] generating AGGRESSIVE request 0 [ SA KE No ID CERTREQ V V V V V ]
[NET] sending packet: from 192.168.0.11[500] to 101.178.197.207[500] (759 bytes)
[NET] received packet: from 101.178.197.207[500] to 192.168.0.11[500] (3478 bytes)
[ENC] parsed AGGRESSIVE response 0 [ SA KE No ID CERT CERT SIG CERTREQ V NAT-D NAT-D V V V V V ]
[IKE] received NAT-T (RFC 3947) vendor ID
[IKE] received DPD vendor ID
[IKE] received XAuth vendor ID
[ENC] received unknown vendor ID: 82:99:03:17:57:a3:60:82:c6:a6:21:de:00:00:00:00
[IKE] received FRAGMENTATION vendor ID
[IKE] received FRAGMENTATION vendor ID
[IKE] received cert request for 'DC=au, DC=com, DC=workplace, CN=workplace-servername-CA'
[IKE] received end entity cert "ST=STATE, L=Hometown, O=workplace Holdings Pty Ltd, CN=fw.workplace.com, E=helpdesk@workplace.com"
[IKE] received issuer cert "DC=au, DC=com, DC=workplace, CN=workplace-servername-CA"
[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048
[CFG]   using certificate "ST=STATE, L=Hometown, O=workplace Holdings Pty Ltd, CN=fw.workplace.com, E=helpdesk@workplace.com"
[CFG]   using trusted ca certificate "DC=au, DC=com, DC=workplace, CN=workplace-servername-CA"
[CFG]   reached self-signed root ca with a path length of 0
[CFG] checking certificate status of "ST=STATE, L=Hometown, O=workplace Holdings Pty Ltd, CN=fw.workplace.com, E=helpdesk@workplace.com"
[CFG]   fetching crl from 'ldap:///CN=workplace-servername-CA,CN=servername,CN=CDP,CN=Public%20Key%20Services,CN=Services,CN=Configuration,DC=workplace,DC=com,DC=au?certificateRevocationList?base?objectClass=cRLDistributionPoint' ...
[LIB] unable to fetch from ldap:///CN=workplace-servername-CA,CN=servername,CN=CDP,CN=Public%20Key%20Services,CN=Services,CN=Configuration,DC=workplace,DC=com,DC=au?certificateRevocationList?base?objectClass=cRLDistributionPoint, no capable fetcher found
[CFG] crl fetching failed
[CFG] certificate status is not available
[IKE] authentication of 'ST=STATE, L=Hometown, O=workplace Holdings Pty Ltd, CN=fw.workplace.com, E=helpdesk@workplace.com' with RSA_EMSA_PKCS1_NULL successful
[IKE] local host is behind NAT, sending keep alives
[IKE] authentication of 'C=AU, ST=STATE, L=Battery Point, O=workplace Holdings Pty Ltd, CN=me@workplace.com, E=me@workplace.com' (myself) successful
[IKE] sending end entity cert "C=AU, ST=STATE, L=Battery Point, O=workplace Holdings Pty Ltd, CN=me@workplace.com, E=me@workplace.com"
[ENC] generating AGGRESSIVE request 0 [ CERT NAT-D NAT-D SIG ]
[ENC] splitting IKE message (2044 bytes) into 2 fragments
[ENC] generating AGGRESSIVE request 0 [ FRAG(1) ]
[ENC] generating AGGRESSIVE request 0 [ FRAG(2/2) ]
[NET] sending packet: from 192.168.0.11[4500] to 101.178.197.207[4500] (1248 bytes)
[NET] sending packet: from 192.168.0.11[4500] to 101.178.197.207[4500] (868 bytes)
[NET] received packet: from 101.178.197.207[4500] to 192.168.0.11[4500] (92 bytes)
[ENC] parsed TRANSACTION request 2072134256 [ HASH CPRQ(X_TYPE X_USER X_PWD) ]
[ENC] generating TRANSACTION response 2072134256 [ HASH CPRP(X_USER X_PWD) ]
[NET] sending packet: from 192.168.0.11[4500] to 101.178.197.207[4500] (108 bytes)
[NET] received packet: from 101.178.197.207[4500] to 192.168.0.11[4500] (92 bytes)
[ENC] parsed TRANSACTION request 3450897709 [ HASH CPS(X_STATUS) ]
[IKE] XAuth authentication of 'myuserid' (myself) successful
[IKE] IKE_SA workplace[3784] established between 192.168.0.11[C=AU, ST=STATE, L=Battery Point, O=workplace Holdings Pty Ltd, CN=me@workplace.com, E=me@workplace.com]...101.178.197.207[ST=STATE, L=Hometown, O=workplace Holdings Pty Ltd, CN=fw.workplace.com, E=helpdesk@workplace.com]
[IKE] scheduling rekeying in 13938s
[IKE] maximum IKE_SA lifetime 15378s
[ENC] generating TRANSACTION response 3450897709 [ HASH CPA(X_STATUS) ]
[NET] sending packet: from 192.168.0.11[4500] to 101.178.197.207[4500] (92 bytes)
[ENC] generating QUICK_MODE request 431372212 [ HASH SA No KE ID ID ]
[NET] sending packet: from 192.168.0.11[4500] to 101.178.197.207[4500] (460 bytes)
[NET] received packet: from 101.178.197.207[4500] to 192.168.0.11[4500] (108 bytes)
[ENC] parsed INFORMATIONAL_V1 request 2485724945 [ HASH D ]
[IKE] received DELETE for IKE_SA workplace[3784]
[IKE] deleting IKE_SA workplace[3784] between 192.168.0.11[C=AU, ST=STATE, L=Battery Point, O=workplace Holdings Pty Ltd, CN=me@workplace.com, E=me@workplace.com]...101.178.197.207[ST=STATE, L=Hometown, O=workplace Holdings Pty Ltd, CN=fw.workplace.com, E=helpdesk@workplace.com]
[IKE] sending cert request for "DC=au, DC=com, DC=workplace, CN=workplace-servername-CA"
[IKE] initiating Aggressive Mode IKE_SA workplace[3785] to 101.178.197.207
initiate failed: establishing CHILD_SA 'vpn' failed
```

I'm told there's nothing meaningful in the fortigate log at that time (not helpful feedback). But an INFORMATIONAL_V1 is met with a DELETE response.  

I can provide level 4 debugging from the system log which contains packet contents it looks like, but I'd rather do that privately if they are useful than publish them.

**Comment by tobiasbrunner** [comment]
> I am running on the assumption that what Fortigate call Phase 2, strongswan calls a CHILD_SA.

Yeah, Phase 1 and 2 are just IKEv1 terms for basically the same thing that IKEv2 calls IKE and Child SA.



> [ENC] generating QUICK_MODE request 431372212 [ HASH SA No KE ID ID ]
> [NET] sending packet: from 192.168.0.11[4500] to 101.178.197.207[4500] (460 bytes)
> [NET] received packet: from 101.178.197.207[4500] to 192.168.0.11[4500] (108 bytes)
> [ENC] parsed INFORMATIONAL_V1 request 2485724945 [ HASH D ]
> [IKE] received DELETE for IKE_SA workplace[3784]

Hm, replying to the Quick Mode request with a Delete for the complete IKE SA is weird. If the proposed traffic selectors or algorithms would be incorrect you'd expect an INVALID_ID_INFORMATION or NO_PROPOSAL_CHOSEN error notify, respectively. The ESP proposal should be fine, however, I don't know about your traffic selectors:

> local_ts  = 0.0.0.0/0
> remote_ts  = 192.168.3.1/32

That doesn't make that much sense and definitely doesn't match what's configured on the other end (with IKEv1 there is generally no narrowing of the traffic selectors, unless you either use strongSwan on both ends or enable the [Cisco Unity extensions](https://docs.strongswan.org/docs/5.9/plugins/unity.html), but the latter only apply to the remote traffic selectors).

And does this setup not use virtual IP addresses (Mode Config in IKEv1)? Because without them, the server will have a hard time routing traffic to clients correctly. You might want to try the following changes:

    connections {
        workplace {
            ...
            vips = 0.0.0.0
            ...
            children {
                vpn {
                    ...
                    local_ts  = dynamic # or just remove it as that's the default and will get replaced with the assigned virtual IP
                    remote_ts  = 0.0.0.0/0 # see comment below
                    ...
                }
            }
        }
    }

With `remote_ts = 0.0.0.0/0` everything will be tunneled to the server (classic roadwarrior scenario). If you don't want that, you could try `192.168.3.1/32` again, but that might not work due to IKEv1's narrowing restrictions. If so, an updown script could be used to customize what's getting tunneled locally after negotiating `0.0.0.0/0` as remote traffic selector.

**Reply by bernd-wechner** [reply]
A genius!

```
$ sudo swanctl --initiate --child vpn --debug 1
[ENC] generating QUICK_MODE request 2055622124 [ HASH SA No KE ID ID ]
[NET] sending packet: from 192.168.0.11[4500] to 101.178.197.207[4500] (460 bytes)
[NET] received packet: from 101.178.197.207[4500] to 192.168.0.11[4500] (444 bytes)
[ENC] parsed QUICK_MODE response 2055622124 [ HASH SA No KE ID ID ]
[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA2_256_128/MODP_2048/NO_EXT_SEQ
[IKE] detected rekeying of CHILD_SA vpn{1}
[IKE] CHILD_SA vpn{3} established with SPIs c473e300_i ddec51fd_o and TS 192.168.100.5/32 === 0.0.0.0/0
initiate completed successfully
```
and this is the alleged IP of the firewalll:

```
$ ping 192.168.3.1
PING 192.168.3.1 (192.168.3.1) 56(84) bytes of data.
64 bytes from 192.168.3.1: icmp_seq=1 ttl=127 time=24.5 ms
64 bytes from 192.168.3.1: icmp_seq=2 ttl=127 time=24.8 ms
64 bytes from 192.168.3.1: icmp_seq=3 ttl=127 time=24.9 ms
^C
--- 192.168.3.1 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2004ms
rtt min/avg/max/mdev = 24.506/24.718/24.883/0.157 ms
```

So now to my work desktop:

```
$ ping 192.168.3.18
PING 192.168.3.18 (192.168.3.18) 56(84) bytes of data.
64 bytes from 192.168.3.18: icmp_seq=1 ttl=127 time=24.2 ms
64 bytes from 192.168.3.18: icmp_seq=2 ttl=127 time=24.9 ms
64 bytes from 192.168.3.18: icmp_seq=3 ttl=127 time=26.1 ms
64 bytes from 192.168.3.18: icmp_seq=4 ttl=127 time=25.0 ms
64 bytes from 192.168.3.18: icmp_seq=5 ttl=127 time=25.3 ms
^C
--- 192.168.3.18 ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 4006ms
rtt min/avg/max/mdev = 24.167/25.096/26.114/0.626 ms
```

And and rdp session fires up fine!

Now all I have to work out is that ancillary thing: I want to ensure I'm still on my DNS not works, and that only traffic to 198.168.3.* gets routed through the tunnel! I'll looking to tools for diagnosing that and configuring it but if you have any tips, I'm ever grateful!

In fact I can confirm that screwed up my DNS (tried a dig, to a local machine and no go, terminated the ike connection and then dig again and go, also this submission failed while ike was up, and will likely succeed now that it's down. Details. But the main part is good. I can work from home on a RDP link! Awesome. The details next week ... keeping my DNS intact and ensuring only those 192.168.3.* targets go through the tunnel and naught else (didn't even need much trickery as the remote LAN uses subnet 3 and I only use 0 and 1 locally so no overlap - though it would a curio how to manage it if there were - I'm guessing that's what virtual IPs are about). I will also inform my IT guys about this:

> Apple clients (iOS and macOS) also support IKEv2 since about 9 years. So that really shouldn't be a reason.

Not sure why they believe they don't if they do. But they may ask me for evidence ;-).

**Reply by tobiasbrunner** [reply, marked as answer]
> $ sudo swanctl --initiate --child vpn --debug 1

Note that due to `start_action = start` there already was an IKE and Child SA established when you called this, so this command basically created a duplicate Child SA (that's why the log only shows a Quick Mode exchange).

> Now all I have to work out is that ancillary thing: I want to ensure I'm still on my DNS not works,

If you want to avoid that DNS servers received from the server are installed, you could disable the _resolve_ plugin (disable `charon.plugins.resolve.load` in [strongswan.conf](https://docs.strongswan.org/docs/5.9/config/strongswanConf.html)).

> and that only traffic to 198.168.3.* gets routed through the tunnel! I'll looking to tools for diagnosing that and configuring it but if you have any tips, I'm ever grateful!

As I mentioned, split-tunneling with IKEv1 isn't possible directly. You could try setting `remote_ts = 198.168.3.0/24`, but if the server doesn't accept that, you have to handle this on the client manually.

Due to the virtual IP, only traffic with the correct source IP (i.e. the virtual IP) is tunneled. You can use that to your advantage and install such a route yourself. If you disable `charon.install_routes` in strongswan.conf, the daemon won't install any routes in table 220. But you can do so yourself in a custom [updown script](https://docs.strongswan.org/docs/5.9/plugins/updown.html):

    connections {
        workplace {
            ...
            children {
                vpn {
                    ...
                    updown  = /path/to/your/script
                    ...
                }
            }
        }
    }

Then in `/path/to/your/script` (doesn't really matter where it's located as long as the daemon can access and execute it, so AppArmor might affect this) add something like this:

```bash
#!/bin/bash

set -o nounset
set -o errexit

REMOTE_SUBNET="198.168.3.0/24"

case "${PLUTO_VERB}" in
    up-client)
        ip route add "${REMOTE_SUBNET}" dev "${PLUTO_INTERFACE}" src "${PLUTO_MY_SOURCEIP}" table 220
        ;;
    down-client)
        ip route del "${REMOTE_SUBNET}" table 220
        ;;
esac
```

> > Apple clients (iOS and macOS) also support IKEv2 since about 9 years. So that really shouldn't be a reason.
> 
> Not sure why they believe they don't if they do. But they may ask me for evidence ;-).

See e.g. Apple's ["Set up a VPN connection on Mac"](https://support.apple.com/en-mn/guide/mac-help/mchlp2963/14.0/mac/14.0) guide, you can go back as far as High Sierra and IKEv2 is mentioned (or refer them to [our docs](https://docs.strongswan.org/docs/5.9/interop/ios.html)).

**Reply by bernd-wechner** [reply]
Some quick observations after a to and fro pile of experimenting. 

But first enormous thanks for you awesome help Tobias! Owe you a lot! I've been learning heaps (about IPSEC).

1. Disabling the resolve plugin worked. My DNS is preserved
2. Did not need the updown script. It turns out `remote_ts = 192.168.3.18/32` works. That is the IP of my work desktop and the only thing we can access on this vpn (we don't have access to anything else on the work LAN and just use that address to start an rdp session on the desktop and from there are like in the office)
3. I'm a bit bothered I missed the `vips =` config because it is described [here](https://docs.strongswan.org/docs/5.9/config/quickstart.html#_roadwarrior_case_with_virtual_ip) and I did read that through, trying to glean what I could (as it's the closest scenario on that page to mine).
4. Specifying one proposal works, specifying a comma separate list of them (all those supported) and it fails. Could be an IKEv1 thing but I was sure I read that a comma separated list worked on IKEv1.
5. I wonder what pools are? The docs are not at all clear and I experimented with using a pool to ensure only traffic to 192.168.3.18/ goes though the tunnel, based on some tute I read ;-). But it seems not to work, and the docs on pools haven't really helped be get what they are and why to use them,
6. It would be nice if we could turn of resolver transfer at the child level rather than at the charon daemon level (globally). But shrug, works for me now, I only have one IPSEC tunnel and VP for now.
7. Should I have something other than `start_action = start`? I tried `trap` and that worked fine too. Not fully on top of why and when to use these. I guess I just want a command to start the VPN, and one to bring it down. It seems initiate works on the child but terminate on the connection.
8. I moved the secrets to the private folder and `include private/*.secret` where they are mode 700. That should sort of be the standard recommendation ;-)

Anyhow, looks like it all works now! I'm just still on a small learning curve. And I'll share the IKEv2 evidence with my helpdesk.

**Reply by tobiasbrunner** [reply]
> 4\. Specifying one proposal works, specifying a comma separate list of them (all those supported) and it fails. Could be an IKEv1 thing but I was sure I read that a comma separated list worked on IKEv1.

Multiple complete proposals separated by commas should work for IKEv1 (but I guess your mileage may vary with different server implementations). What's not supported with IKEv1 is listing multiple algorithms in a single proposal.

> 5\. I wonder what pools are?

A [server setting](https://docs.strongswan.org/docs/5.9/features/vip.html#_responder_configuration) to define IP address pools from which virtual IPs are assigned to clients.

> 7\. Should I have something other than `start_action = start`? I tried `trap` and that worked fine too. Not fully on top of why and when to use these.

Try [reading this](https://docs.strongswan.org/docs/5.9/howtos/introduction.html#_invocation_and_maintenance).

> 8\. I moved the secrets to the private folder and `include private/*.secret` where they are mode 700. That should sort of be the standard recommendation ;-)

The `private` directory is intended for private keys. But sure, you could use it that way too. Note that `swanctl.conf` is owned by root:root and has mode 640 by default (not sure if any distro changes this).
